text/0000-at-least-one-feature.md

```diff
@@ -0,0 +1,296 @@
+
+In step two, we could "ret-con" the empty feature set with the `default` feature.
+But this is a trick that can only be pulled once.
+The second time around, we already have a default feature; we are out of luck.
+
+The previous attempts attempted to make new default features, or migrate the requested feature sets.
+But that is complex.
+There is exactly a simpler solution: simply require that *some* feature always be depended-upon.
+
+To see why this works, it helps to first see that step 2 was *already* broken.
+Here's the thing, even though there previously were not any features, that *doesn't* mean there were not any `no-default-features = true` users!
+Sure, it wouldn't do anything for crate with new features, but one can still use it.
+Then when just `bar` is added, we already have a problem, because the `default` feature will no "catch" all the existing users ---
+the mischievous users that were already using `no-default-features = true` will have their code broken!
+
+This brings us to the heart of the problem.
+So long as users are depending on "something", we can be careful to make sure those features keep their meaning.
+But when users are depending on nothing at all with `no-default-features = true` and an empty feature set, we have nothing to "hook into".
+The simple solution is just to rule out that problem entirely!
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Packages with
+```toml
+[package]
+empty-features = false
+```
+are easier to maintain!
+You don't need to worry about the `no-default-features = true, features = []` case anymore.
+You can be sure that all consumers must dependent on either `default` or a regular named feature.
+
+Whenever you want to make existing functionality more conditional, simply extend the set of features the code relies on with a new feature, and ensure existing features depend on that feature.
+
+For example:
+```rust
+fn my_fun_that_allocates() { .. }
+
+#[cfg(all(feature = "foo", feature = "bar"))]
+fn my_weird_fun_that_allocates() { .. }
+```
+becomes:
+```rust
+#[cfg(feature = "baz")]
+fn my_fun_that_allocates() { .. }
+
+#[cfg(all(feature = "foo", feature = "bar", feature = "baz"))]
+fn my_weird_fun_that_allocates() { .. }
+```
+
+And the corresponding `Cargo.toml`:
+```toml
+[features]
+default = ["foo"]
+bar = ["foo"]
+```
+becomes:
+```toml
+[features]
+default = ["foo", "baz"]
+foo = ["baz"]
+bar = ["foo"] # no need to add "baz" because "bar" picks it up
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Depending on a
+```toml
+[package]
+empty-features = false
+```
+crate with an empty feature set is disallowed and invalidates the solution.
+The `default` feature counts as a member of that set when `default-features = true`.
+
+The solver shall avoid such solutions (so as not break old versions of libraries without `at-least-one-feature` being "discoverable").
```

Comment:
How? Is it supposed to just pick a feature and enable that?

I am also confused by the paragraph on why adding `empty-features = false` is not a breaking change, which is probably related to my confusion about this sentence.

Reply:
I just meant if no feature is enable than the plan is invalid, and it must throw it out. (If all plans don't enable any feature for a `empty-features = false` crate, then no plans are valid)

> Is it supposed to just pick a feature and enable that?

Hehe I didn't even think of that. I suppose it is quite easy to take any such invalid plan and make it valid in this way, but yeah, ew.

Reply:
Oh I get it, so the reason this is semver compatible to add is that cargo will refuse to use the newer version because that would violate this condition -- so it is stuck with the old version?

That makes sense, but could be explained in a bit more detail to make it more clear.